### PR TITLE
invalid engines

### PR DESCRIPTION
**Pull request by ABaldwinHunter**
@codeclimate/review 

This PR adds check for invalid engine names to `codeclimate validate-config`

Its implementation is to create a `validate` subcommand under `enges` (which means that users can also run `codeclimate engines:validate`) which I kind of like.

Feedback welcome.

### CODE REVIEW

File: lib/cc/cli/engines.rb [hunk 1]
```diff
@@ -9,6 +9,7 @@ module Engines
       autoload :Install, "cc/cli/engines/install"
       autoload :List, "cc/cli/engines/list"
       autoload :Remove, "cc/cli/engines/remove"
+      autoload :Validate, "cc/cli/engines/validate"
```
Comment (pbrisbin) [inline lib/cc/cli/engines.rb]: I thought we wanted this to be incorporated into the existing `validate-config` command. Might be a question for the team, both ways make some sense.

File: .codeclimate.yml [hunk 2]
```diff
@@ -4,6 +4,8 @@ engines:
     enabled: true
   bundler-audit:
     enabled: true
+  hippo:
+    enabled: true
```
Comment (pbrisbin) [inline .codeclimate.yml]: accidental?

**Comment by brynary** [comment]
Seems like this should be part of the `codeclimate validate-config` experience. Pat -- What's your thinking on why a separate subcommand may make sense?

File: lib/cc/cli/engines/validate.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+require "cc/analyzer"
+
+module CC
+  module CLI
+    module Engines
+      class Validate < EngineCommand
+        def run
+          if invalid_engines.any?
+            warn
+            true
```
Comment (pbrisbin) [inline lib/cc/cli/engines/validate.rb]: We would want this command to exit non-zero to signify failure in this case right? I'm not sure the mechanism commands have for controlling that, but if it's returning `true` here, that seems backwards.

**Comment by pbrisbin** [comment]
@brynary I agree with you and made that comment [myself](https://github.com/codeclimate/codeclimate/pull/75#discussion_r35993171)

**Comment by pbrisbin** [comment]
@brynary scratch that -- I forgot I said both may make some sense earlier... The sense I see in this way is:

- It still is part of `validate-config` experience, `engines:validate` is a bonus
- I wouldn't normally say it was worth it except this implementation is quite small and actually easier this way than the other (because of our current inheritance-based architecture)

Up to me: I'd refactor the inheritance in the `Engines` commands to use composition for the interaction with the config that needs to now be shared. Then I'd use that shared object to implement this only in the `validate-config` command.

**Comment by ABaldwinHunter** [comment]
@brynary @pbrisbin that sounds pretty good.

As an FYI or reminder for consideration- I think the reasons we went with inheritance originally was that at least half of the current collection of engines subcommands (`remove`, `enable`, `disable`) reference  only one engine. (engines:enable <x>, engines:remove <x>). These subcommands are [initialized] (https://github.com/codeclimate/codeclimate/blob/master/lib/cc/cli/engines/engine_command.rb#L11) with an `@engine_name` variable that reduces argument passing. 

That structure might not make sense as our need for `EngineRegistry` and engines stuff in other places expands, but wanted to raise that thought.

File: lib/cc/cli/engines/validate.rb (hunk 3 above)
Comment (brynary) [inline lib/cc/cli/engines/validate.rb]: I don't think we have that exit code behavior currently, so if so we should do that separately. Probably want a CommandResult object so we don't litter exit calls in the classes

**Comment by brynary** [comment]
Ok. Yeah I thinkI agree with Pat's proposal

**Comment by ABaldwinHunter** [comment]
@brynary @pbrisbin okay that implementation sounds good to me. Thanks for the discussion.

**Comment by ABaldwinHunter** [comment]
See https://github.com/codeclimate/codeclimate/pull/78/files for composition implementation.

**Pull request closed without merging**